lib/trie/database.go

```diff
@@ -61,241 +61,270 @@ func (t *Trie) store(db chaindb.Batch, curr Node) error {
-// it writes these in a batch operation.
+// DeleteFromDB deletes a value from the trie at the key given.
+// It writes the updated nodes from the changed node up to the root node
+// to the database in a batch operation.
 func (t *Trie) DeleteFromDB(db chaindb.Database, key []byte) error {
 	t.Delete(key)
 	return t.WriteDirty(db)
 }
 
-// ClearPrefixFromDB deletes all keys with the given prefix from the trie
-// and writes the updated nodes the database. Since it needs to write all
-//  the nodes from the changed node up to the root, it writes these
+// ClearPrefixFromDB deletes all nodes with keys starting the given prefix
+// from the trie. It writes the updated nodes from the changed node up to
+// the root node to the database in a batch operation.
 // in a batch operation.
 func (t *Trie) ClearPrefixFromDB(db chaindb.Database, prefix []byte) error {
 	t.ClearPrefix(prefix)
 	return t.WriteDirty(db)
 }
 
-// GetFromDB retrieves a value from the trie using the database.
+// GetFromDB retrieves a value at the given key from the trie using the database.
 // It recursively descends into the trie using the database starting
 // from the root node until it reaches the node with the given key.
 // It then reads the value from the database.
-func GetFromDB(db chaindb.Database, root common.Hash, key []byte) ([]byte, error) {
-	if root == EmptyHash {
+func GetFromDB(db chaindb.Database, rootHash common.Hash, key []byte) (
+	value []byte, err error) {
+	if rootHash == EmptyHash {
 		return nil, nil
 	}
 
 	k := codec.KeyLEToNibbles(key)
 
-	enc, err := db.Get(root[:])
+	encodedRootNode, err := db.Get(rootHash[:])
 	if err != nil {
-		return nil, fmt.Errorf("failed to find root key=%s: %w", root, err)
+		return nil, fmt.Errorf("cannot find root hash key 0x%x: %w", rootHash, err)
 	}
 
-	rootNode, err := node.Decode(bytes.NewReader(enc))
+	reader := bytes.NewReader(encodedRootNode)
+	rootNode, err := node.Decode(reader)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot decode root node: %w", err)
 	}
 
 	return getFromDB(db, rootNode, k)
 }
 
-func getFromDB(db chaindb.Database, parent Node, key []byte) ([]byte, error) {
-	var value []byte
-
-	switch p := parent.(type) {
-	case *node.Branch:
-		length := lenCommonPrefix(p.Key, key)
-
-		// found the value at this node
-		if bytes.Equal(p.Key, key) || len(key) == 0 {
-			return p.Value, nil
-		}
-
-		// did not find value
-		if bytes.Equal(p.Key[:length], key) && len(key) < len(p.Key) {
-			return nil, nil
-		}
-
-		if p.Children[key[length]] == nil {
-			return nil, nil
+// getFromDB recursively searches through the trie and database
+// for the value corresponding to a key.
+// Note it does not copy the value so modifying the value bytes
+// slice will modify the value of the node in the trie.
+func getFromDB(db chaindb.Database, n Node, key []byte) (
+	value []byte, err error) {
+	// if parent == nil {
+	// 	return nil, nil
+	// }
```

Comment:
```suggestion
```

Reply:
Yeah left it to try it out, I ran the node in production and never encountered a panic. I'll assume this never happen :wink: